### PR TITLE
[browser][non-icu] `HybridGlobalization` compare

### PR DESCRIPTION
**Pull request by ilonatommy**
Implements a chunk of web-api based globalization. Is a part of HybridGlobalization feature and contributes to https://github.com/dotnet/runtime/issues/79989.

Old, icu-based private API: GlobalizationNative_CompareString

New, non-icu private API: CompareString.

Affected public API (see: tests in CompareInfoTests.Compare.cs):
- CompareInfo.Compare,
- String.Compare,
- String.Equals.

All changes in behavior are listed in `docs\design\features\hybrid-globalization.md`

|    **Test name**    | **time ICU4C [ms]** | **time HG [ms]** | **increase by [times]** |
|-:|-:|-:|-:|
|                 String, String Compare |     0.0448 |  5.029 | 112
|                  String, String Equals |     0.0450 | 4.9587 | 110
|            String, CompareInfo Compare |     0.0451 |  4.9238 | 109

see final performance here: https://github.com/dotnet/runtime/pull/85098
cc @SamMonoRT 

Performance:
most of the processing time is taken by `mono_wasm_compare_string` (~60-70%) while ~50% is used for TextDecoder.decode and only up to 20% for localeCompare.

### CODE REVIEW

**Comment by ilonatommy** [comment]
Failures: Blazor WasmBuildTests and AOT timeouts are not connected.

File: docs/design/features/hybrid-globalization.md [hunk 1]
```diff
@@ -18,3 +18,157 @@ Affected public APIs:
...
+
+The subset of ignored symbols is limited to the symbols ignored by `string1.localeCompare(string2, locale, { ignorePunctuation: true })`. E.g. currency symbols, & are not ignored
+
+```
+let hiraganaAccent = `${String.fromCharCode(12353)} á`                     // %u3041 = ぁ
+let katakanaNoAccent = `${String.fromCharCode(12449)} a`                   // %u30A1 = ァ
+hiraganaBig.localeCompare(katakanaSmall, "en-US", { sensitivity: "base" }) // 0; base: a ≠ b, a = á, a = A
+```
+
+- List of all `CompareOptions` combinations always throwing `PlatformNotSupportedException`:
```
Comment (kg) [inline docs/design/features/hybrid-globalization.md:130]: In the future, could we use attributes or a compile time analyzer to make builds fail when the user's code attempts to use these combinations? (Obviously not in this PR)

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.WebAssembly.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+using System.Runtime.CompilerServices;
+
+namespace System.Globalization
+{
+    internal static unsafe class CompareInfoInterop
+    {
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        internal static extern unsafe int CompareStringJS(out string exceptionMessage, in string culture, char* str1, int str1Len, char* str2, int str2Len, global::System.Globalization.CompareOptions options);
+    }
+    public partial class CompareInfo
```
Comment (kg) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.WebAssembly.cs:15]: new line here Ideally

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.WebAssembly.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            if (CompareOptionsNotSupportedForCulture(options, cultureName))
+                throw new PlatformNotSupportedException(GetPNSEForCulture(options, cultureName));
+
+            string exceptionMessage;
+            int cmpResult;
+            fixed (char* pString1 = &MemoryMarshal.GetReference(string1))
+            fixed (char* pString2 = &MemoryMarshal.GetReference(string2))
+            {
+                // it hovers when I call the below:
+                // investigate: how far does it get? log stuff in JS
```
Comment (kg) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.WebAssembly.cs:39]: what do these comments mean?

Comment (ilonatommy) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.WebAssembly.cs:39]: Honestly, no idea. Some old trash, removing.

File: src/mono/wasm/runtime/net6-legacy/method-calls.ts [hunk 4]
```diff
@@ -315,7 +316,7 @@ export function mono_wasm_change_case_invariant(exceptionMessage: Int32Ptr, src:
     }
 }
 
-export function mono_wasm_change_case(exceptionMessage: Int32Ptr, culture: MonoStringRef, src: number, srcLength: number, dst: number, destLength: number, toUpper: number) : void{
+export function mono_wasm_change_case(exceptionMessage: Int32Ptr, culture: MonoStringRef, src: number, srcLength: number, dst: number, destLength: number, toUpper: boolean) : void{
```
Comment (kg) [inline src/mono/wasm/runtime/net6-legacy/method-calls.ts:319]: boolean -> number i think? unless something changed

File: src/mono/wasm/runtime/net6-legacy/method-calls.ts [hunk 5]
```diff
@@ -337,18 +338,26 @@ export function mono_wasm_change_case(exceptionMessage: Int32Ptr, culture: MonoS
...
-        string += String.fromCharCode(view[i]);
-    return string;
+export function mono_wasm_compare_string(exceptionMessage: Int32Ptr, culture: MonoStringRef, str1: number, str1Length: number, str2: number, str2Length: number, options: number) : number{
+    const cultureRoot = mono_wasm_new_external_root<MonoString>(culture);
+    try{
+        const cultureName = conv_string_root(cultureRoot);
+        const string1  = get_uft16_string(str1, str1Length);
+        const string2 = get_uft16_string(str2, str2Length);
+        const casePicker = (options & 0x1f);
+        const locale = (cultureName && cultureName?.trim()) ? cultureName : undefined;
```
Comment (kg) [inline src/mono/wasm/runtime/net6-legacy/method-calls.ts:348]: this logic seems a little weird to me, if we're trimming the name shouldn't we be using the trimmed name here? or is this just meant to be IsNullOrWhiteSpace

File: src/mono/wasm/runtime/net6-legacy/method-calls.ts (hunk 4 above)
Comment (ilonatommy) [reply src/mono/wasm/runtime/net6-legacy/method-calls.ts:319]: Merge error

File: src/mono/wasm/runtime/net6-legacy/method-calls.ts (hunk 5 above)
Comment (ilonatommy) [reply src/mono/wasm/runtime/net6-legacy/method-calls.ts:348]: It was meant to be IsNullOrWhitespace but now I think doing `if (cultureName)` check will do good as well. I will simplify it

File: src/mono/wasm/runtime/net6-legacy/globalization-helper.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    const view = new Uint16Array(Module.HEAPU16.buffer, ptr, length);
+    let string = "";
+    for (let i = 0; i < length; i++)
+        string += String.fromCharCode(view[i]);
+    return string;
+}
+
+export function pass_exception_details(ex: any, exceptionMessage: Int32Ptr){
+    const exceptionJsString = ex.message + "\n" + ex.stack;
+    const exceptionRoot = mono_wasm_new_root<MonoString>();
```
[the remark this answers is not in the document]
Comment (ilonatommy) [reply src/mono/wasm/runtime/net6-legacy/globalization-helper.ts:20]: I like this idea, applying

**Review by kg — APPROVED** [review]

File: src/libraries/System.Globalization/tests/CompareInfo/CompareInfoTests.Compare.cs [hunk 7]
```diff
@@ -13,6 +13,16 @@ public class CompareInfoCompareTests
         private static CompareInfo s_currentCompare = CultureInfo.CurrentCulture.CompareInfo;
         private static CompareInfo s_hungarianCompare = new CultureInfo("hu-HU").CompareInfo;
         private static CompareInfo s_turkishCompare = new CultureInfo("tr-TR").CompareInfo;
+        private static CompareInfo s_japaneseCompare = new CultureInfo("ja-JP").CompareInfo;
+        private static CompareOptions supportedIgnoreNonSpaceOption =
+            PlatformDetection.IsHybridGlobalization ?
```
Comment (mkhamoyan) [inline src/libraries/System.Globalization/tests/CompareInfo/CompareInfoTests.Compare.cs:18]: Here will be better to mention for Wasi/Browser. For now there is no implementation for OSX platforms but once native functions will be added for hybrid mode this needs to be changed.

File: src/libraries/System.Globalization/tests/CompareInfo/CompareInfoTests.Compare.cs [hunk 8]
```diff
@@ -50,68 +65,72 @@ public static IEnumerable<object[]> Compare_Kana_TestData()
...
-            yield return new object[] { s_invariantCompare, "\u30FC", "\uFF70", ignoreKanaIgnoreWidthIgnoreCase, 0 };
-            yield return new object[] { s_invariantCompare, "\u30FC", "\uFF0D", ignoreKanaIgnoreWidthIgnoreCase, 1 };
-            yield return new object[] { s_invariantCompare, "\u30FC", "\u30FC", ignoreKanaIgnoreWidthIgnoreCase, 0 };
-            yield return new object[] { s_invariantCompare, "\u30FC", "\u2015", ignoreKanaIgnoreWidthIgnoreCase, 1 };
-            yield return new object[] { s_invariantCompare, "\u30FC", "\u2010", ignoreKanaIgnoreWidthIgnoreCase, 1 };
-
-            yield return new object[] { s_invariantCompare, "/", "\uFF0F", ignoreKanaIgnoreWidthIgnoreCase, 0 };
-            yield return new object[] { s_invariantCompare, "'", "\uFF07", ignoreKanaIgnoreWidthIgnoreCase, 0 };
-            yield return new object[] { s_invariantCompare, "\"", "\uFF02", ignoreKanaIgnoreWidthIgnoreCase, 0 };
+            // PlatformDetection.IsHybridGlobalization does not support IgnoreWidth
```
Comment (mkhamoyan) [inline src/libraries/System.Globalization/tests/CompareInfo/CompareInfoTests.Compare.cs:68]: Here also in comments better to add for Wasi/Browser platforms.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.WebAssembly.cs [hunk 9]
```diff
@@ -0,0 +1,67 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+namespace System.Globalization
+{
+    internal static unsafe class CompareInfoInterop
```
Comment (mkhamoyan) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.WebAssembly.cs:10]: Wouldn't it better to separate JS calls like it is done now for Unmanaged code calls https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/Interop/Interop.Casing.cs ?

File: src/libraries/System.Globalization/tests/CompareInfo/CompareInfoTests.Compare.cs (hunk 7 above)
Comment (ilonatommy) [reply src/libraries/System.Globalization/tests/CompareInfo/CompareInfoTests.Compare.cs:18]: Thanks, changing to `IsHybridGlobalizationOnWasm`.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.WebAssembly.cs (hunk 9 above)
Comment (ilonatommy) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.WebAssembly.cs:10]: You're right, we even have a special `Browser` directory there where I can place the interop calls.

**Comment by ilonatommy** [comment]
Failures not related. Merging to move forward. If there are any new ideas later, feel free to let me know and we will solve them in a follow-up, cc @pavelsavara, @radical.

**Pull request merged**